ボウリング（オブジェクト指向版）の課題を提出いたします

ham-cap opened the pull request:
オブジェクト指向版のボウリングのプログラムを提出いたします。
よろしくお願いいたします。

ham-cap committed:
First commit

ham-cap committed:
shot,frame,gameの各クラスを作成・それぞれのテストを作成

ham-cap committed:
実際にスコア計算を行うbowling.rbを作成・Frameクラスのコンストラクタの処理を修正・Frameクラス修正に合わせてテストも修正・rubocopのoffencesを自動修正

・Frameクラスのコンストラクタで10フレーム目の3投目も引数で渡せるようにしていたが、現在のカウント方法だと不要なため削除した

ham-cap committed:
game.rbのコンストラクタの処理をメソッドに切り分けた・bowling.rb内の処理の一部をShotとFrameのクラスメソッド化した・game_spec.rbを一部修正した

・Gameクラスのコンストラクタ内の条件分岐が複雑すぎて、rubocopのMetrics/PerceivedComplexity: で引っかかったため、分岐の判定処理をメソッドに切り分けた。
・bowling.rb内で行う処理の一部について、Shot・Frameの各クラスと関連が強いと判断したため、各クラスのインスタンスの配列を生成する処理をそれぞれのクラスメソッド化した。
・Gameクラスのテストでdiscribe内の記述が長すぎたためメソッド定義を外に出した。

ham-cap committed:
bowling.rb内の書き方を修正・ShotとFrameのクラスメソッドの名称を変更・rubocopで引っかかったインデントを自動修正

・bowling.rb内でGameのインスタンスを生成する箇所の記述がわかりにくかったため修正した
・ShotとFrameにそれぞれ追加したクラスメソッドの名称が冗長だったため変更した

yoshitsugu commented on 07.bowling_object/Rakefile:
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+require 'rspec/core/rake_task'
```
外部gemを参照するときは、Gemfileを作っておくと確認もしやすいです〜

yoshitsugu commented on 07.bowling_object/bowling.rb:
```diff
@@ -0,0 +1,23 @@
+end
+
+shots = Shot.create_array(@scores)
+frames = Frame.create_array(shots)
```
ここで引数のエラーになってしまうんですが、実行できています？
```
frame.rb:15:in `create_array': wrong number of arguments (given 1, expected 0) (ArgumentError)
from bowling.rb:21:in `<main>'
```

ham-cap replied:
ご指摘ありがとうございます🙏
確認してみたところ、うまく動いていませんでした。
申し訳ありません🙇‍♂️
オブジェクト指向版ではないほうのボウリングの課題の時に今回の実行ファイルと同名の`bowling.rb`のパスを通していたことを失念しており、動作確認の際にたまにそちらを呼び出してしまっていたようで、うまく動いているように見えてしまっていました💧
もう一度見直して問題のあった箇所を修正のうえプッシュしましたのでご確認お願いします🙇‍♂️

yoshitsugu commented on 07.bowling_object/game.rb:
```diff
@@ -0,0 +1,37 @@
+  end
+
+  def add_bonus_for_strike_and_spare(frame, index)
+    @scores_of_each_frame << if frame.first_shot.score == 10 && frames[index + 1].first_shot.score != 10
```
できるだけ状態を変化させるメソッドは少なくしましょう。
ここは `add_bonus_for_strike_and_spare` の呼び出し側で

```ruby
 @scores_of_each_frame <<  add_bonus_for_strike_and_spare(frame, index)
```

としたほうがよいかなと思います。
なぜかというと、あちこちから状態が変更されると、何かバグが発生したときに、どこで状態が変わってそうなっているのかよくわからなくなってしまいがちだからです。

yoshitsugu commented on 07.bowling_object/shot.rb:
```diff
@@ -0,0 +1,20 @@
+  end
+
+  def self.create_array(scores)
+    @shots = []
```
インスタンス変数、クラス変数は必要な箇所でのみ使うようにするとよいです。
今回くらいであればあまり関係ないですが、厳密にいうとインスタンスが生存している間インスタンス変数も生存しています。クラス変数であればクラスが定義されている間ずっとなので、最初に実行されてから実行がおわるまでずっと保持しつづけられます。ローカル変数のほうが生存期間が短かくてメモリ効率がよいです。
また、今回であれば単に

```ruby
 scores.map { |score| Shot.new(score) }
```

で十分であり、そもそもメソッド化する必要もないかもです。

yoshitsugu commented on 07.bowling_object/frame.rb:
```diff
@@ -0,0 +1,21 @@
+  end
+
+  def self.create_array
+    @frames = []
+    Shot.create_shots_array(@scores).each_slice(2).each do |a_pair_of_shots|
+      @frames << Frame.new(a_pair_of_shots[0], a_pair_of_shots[1])
+    end
```
Shot.create_shots_arrayと同様、こちらも `map` で十分かなと思います。

yoshitsugu reviewed (commented):
コメントしました！

yoshitsugu commented on 07.bowling_object/frame_spec.rb:
```diff
@@ -0,0 +1,19 @@
+  let(:strike_shot) { Shot.new('X') }
+  let(:normal_frame) { Frame.new(first_shot, second_shot) }
+
+  it 'frameオブジェクトが生成されているか' do
```
これはあまりテストする意味がない気もします。
16行目のテストでオブジェクトがちゃんとできてなかったらエラーになりますしね。

yoshitsugu commented on 07.bowling_object/game_spec.rb:
```diff
@@ -0,0 +1,55 @@
+require 'byebug'
+
+def make_a_frame(inputed_line)
+  @scores = []
```
他のところでも書いたとおり、できるだけインスタンス変数ではなく、ローカル変数を使いましょう

yoshitsugu commented on 07.bowling_object/game_spec.rb:
```diff
@@ -0,0 +1,55 @@
+  let(:perfect_game) { Game.new(make_a_frame(@input2)) }
+  let(:normal_game) { Game.new(make_a_frame(@input3)) }
+
+  it 'Gameオブジェクトが生成されているか' do
```
frame_specと同じくこれはあまり意味ないかもです。

yoshitsugu commented on 07.bowling_object/shot_spec.rb:
```diff
@@ -0,0 +1,20 @@
+  let(:strike_shot) { Shot.new('X') }
+  let(:not_strike_shot) { Shot.new('9') }
+
+  it 'Shotオブジェクトが生成されているか' do
```
同上です。

yoshitsugu reviewed (commented):
コメント追加です。

ham-cap committed:
Gemfileを作成

ham-cap committed:
FrameとShotのクラスメソッドを廃止・bowling.rbの処理を修正・一部のインスタンス変数をローカル変数へ修正・Shotのコンストラクタでの処理を修正

・FrameとShotのcreate_arrayメソッドでは複雑な処理をしないため、メソッドとしては切り出さずにbowling.rbで同様の処理を行うこととした。
・使用範囲が狭くインスタンス変数である必要のない変数をローカル変数へ修正した。
・Shotのコンストラクタでstringの'X'をintegerの10へ変換する処理を行っていたが、bowling.rbでも同様の処理を行なっており、その後の要素数の調整もbowling.rb側で行なっていることから、Shotのコンストラクタでは単にstringをintegerへの変換のみを行うこととした。

ham-cap committed:
不要なテストを削除し、bowling.rbでの処理修正に合わせてテストのロジックも修正した

yoshitsugu commented on 07.bowling_object/bowling.rb:
```diff
@@ -0,0 +1,27 @@
+  scores << 0
+end
+
+shots = scores.map{|score| Shot.new(score)}
```
mapのあとの `{` の前後って空白ないとRubocopに怒られませんでしたっけ。
rubocop-fjordの設定を再度見直してみてもらってよいでしょうか。
https://bootcamp.fjord.jp/pages/292

ham-cap replied:
すみません。こちら再提出の際にrubocopを通し忘れていたためです。
失礼しました🙇‍♂️

yoshitsugu commented on 07.bowling_object/game.rb:
```diff
@@ -0,0 +1,36 @@
+  end
+
+  def score
+    scores_of_each_frame = []
+    @frames.each_with_index do |frame, index|
+      if final_frame?(index)
+        scores_of_each_frame << frame.score
+      else
+        scores_of_each_frame << add_bonus_for_strike_and_spare(frame, index)
+      end
+    end
+    scores_of_each_frame.sum
```
どうせ最後にsumするだけなら、each_with_indexの中で `+= ` で足し込んでいってもいいのではと思いました。

yoshitsugu commented on 07.bowling_object/game_spec.rb:
```diff
@@ -0,0 +1,51 @@
+  end
+
+  shots = scores.map{|score| Shot.new(score)}
+  frames = []
+  shots.each_slice(2).to_a.map{|pair_of_shots| frames << Frame.new(pair_of_shots[0], pair_of_shots[1])}
+  frames
```
mapの使い方、再度確認しましょう！
```ruby
array1 = [1,2,3]
array2 = array1.map { |a| a + 1 }
array2
#=> [2,3,4]
```
mapはブロックの返り値が配列の要素になるので、その中で `frames <<` みたいに一個ずつつめる必要はありません。

ham-cap replied:
承知しました。
よく考えると`to_a`も不要のようだったので削除しました。

yoshitsugu reviewed (commented):
コメントしました！

ham-cap committed:
Gameクラスのscoreメソッド内の処理を修正・gameのテストで冗長だった箇所を修正・rubocopによる修正を適用

・Gameクラスのscoreメソッドは各フレームの特典を配列にして最後にsumで合計していたが、繰り返し処理の段階で加算していく方法に変更した。
・gameのテスト内でmapを使用している箇所について、ブロック内でFrameのインスタンスを生成し配列に追加していく処理を行っていたが、mapの返り値は配列の要素として返るため、ブロック内で配列に追加する必要はなかった。

ham-cap committed:
make_a_frameメソッド内の処理を一部修正

・Frameのインスタンスを配列にする際にメソッドチェーン内でto_aを使用していたが、不要だったため削除。

yoshitsugu commented on 07.bowling_object/bowling.rb:
```diff
@@ -0,0 +1,25 @@
+
+scores << 0 if scores.size.odd?
+
+shots = scores.map { |score| Shot.new(score) }
+frames = shots.each_slice(2) { |pair_of_shots| Frame.new(pair_of_shots[0], pair_of_shots[1]) }
+game = Game.new(frames)
+puts game.score
```
オブジェクト指向では、オブジェクトに何かのメッセージを送ることで、オブジェクトがよしなにやってくれる、というところが特徴としてあります。これにより処理の内部を知らずに、使う側はインターフェースだけ知ればよい、ということになります。
ここでは、 `Shot` と `Frame` と `Game` をこのように組み立てる、という知識がここに漏れてしまっているので、ややオブジェクト指向的でないように感じました。
`Game` のクラスメソッドなど、別のクラスに隠蔽してしまってもよいかと思います。

ham-cap replied:
おっしゃっているのはポリモーフィズムやダックタイピングのことかなと思いましたので、参考書籍等でもう一度確認してみたところ、疑問が出てきたので教えてください。

今回、`Game`・`Frame`・`Shot`はそれぞれ独立したクラスとして定義していて継承を用いていないため、この点がまずオブジェクト指向の要素を満たしていないという理解で大丈夫でしょうか。
その場合にちょっと疑問なのは、例えば今回の場合`score`メソッドをスーパークラスに定義したとしてもおそらく中身は空になると思います。当然、継承した具体の三つのクラスがそれぞれのクラス内部でオーバーライドすることになりますが、そうなるとそもそもスーパークラスを用意する意味ってあるのでしょうか？🤔
各クラスで同名のメソッドさえ用意しておけば同じことなのではと思えるのですが。
クラスが増えた時のことを考えるにしても、新しいクラスでメソッド定義を忘れて使用した時にエラーになるか`nil`が返るかの違いという気もします。
まずこれがそもそもの疑問です。

それと実装について確認なのですが、コメントでいただいた
> `Game`のクラスメソッドなど、別のクラスに隠蔽してしまってもよいかと思います。

という部分で、ここでいう「隠蔽する」とは「各クラスを使用する側（bowling.rb）から`Game`のオブジェクトが生成されるまでのプロセスを隠蔽する」という意味で受け取ったのですが大丈夫でしょうか。
仮にそうだとした場合、ご提案いただいたように`Game`のクラスメソッドか、あるいはファクトリーパターンのように各クラスのオブジェクト生成を担う専門のFactoryクラスを用意して、スコアの配列さえ渡せばその値に基づいた`Game`クラスのオブジェクトが生成されて返って来て、そのオブジェクトに`score`メソッドでメッセージを送るとトータルスコアが分かる、という流れで考えてみたのですがいかがでしょうか。

おかしい部分があれば教えてください。

yoshitsugu replied:
> 今回、Game・Frame・Shotはそれぞれ独立したクラスとして定義していて継承を用いていないため、この点がまずオブジェクト指向の要素を満たしていないという理解で大丈夫でしょうか。

いえ、継承は特に関係ないですね。オブジェクト指向をすすめると、オブジェクトにメッセージを送るのみのでオブジェクト側で処理をよしなにやってくれる、という形になっていきます。
こうすることで、処理の内部実装が多少変わったとしてもそのオブジェクトのみに変更を閉じることができ、変更に強いプログラムになる、というのがオブジェクト指向のよさでしたよね。
そういう点でこのコードを眺めてみると、「2のShotの組がFrameになること」や「Frameの配列がGameに必要なこと」などオブジェクトを作る上での前提知識を利用する側が結構深く知っている必要があります。これではオブジェクト指向のよさである処理の隠蔽が中途半端になっていないか、というのが今回の論点です。

> ここでいう「隠蔽する」とは「各クラスを使用する側（bowling.rb）からGameのオブジェクトが生成されるまでのプロセスを隠蔽する」という意味で受け取ったのですが大丈夫でしょうか。

はい。これはあっています。
ファクトリーパターンで解決してもいいですし、単純にGameクラスが文字列からのオブジェクト生成方法を知っている、としてもいいかと思います。

ham-cap replied:
ありがとうございます！承知しました！
それらを踏まえて直してみますので少しお時間ください🙏

yoshitsugu commented on 07.bowling_object/game.rb:
```diff
@@ -0,0 +1,37 @@
+# frozen_string_literal: true
+
+class Game
+  attr_reader :frames
```
`attr_reader` はクラスの外部から参照される場合に必要ですが、 `frames` は外部から参照されていなさそうに見えます。いかがでしょうか？

yoshitsugu reviewed (commented):
コメントしました！

ham-cap committed:
attr_readerを削除

・@framesはGameクラス外部から参照されていないため削除した。

ham-cap committed:
Gameオブジェクトの生成プロセスを同クラスのクラスメソッド化

・Gameオブジェクトの生成プロセスは全てbowling.rbにおいて実行していたが、bowling.rb側はShotやFrameオブジェクトの生成方法まで知っている必要はないため、Gameのクラスメソッドにスコアの配列を渡せば全て生成されるように修正した。

ham-cap commented:
修正しました！

yoshitsugu commented on 07.bowling_object/bowling.rb:
```diff
@@ -0,0 +1,22 @@
+require_relative 'shot'
+
+inputed_line = ARGV[0]
+marks = inputed_line.split(',')
+
+scores = []
+marks.each do |mark|
+  if mark == 'X'
+    scores << 10
+    scores << 0
+  else
+    scores << mark
+  end
+end
+
+scores << 0 if scores.size.odd?
```
この辺の処理も `Game` が知っていることとして、文字列だけ渡したら実行してくれる形でもよいとは思いますが、このままでもOKです。

yoshitsugu reviewed (approved):
OKです！
